Environment Set-Up (#56)
hytest-org/hytest: environment_set_up/QuickStart-HPC.md

# Quick-Start for HPC

This document will help you set up the correct computing environment on a compute
server, and access the notebooks on your local desktop.

## 1) Configure the software environment

This should only need to be done once. After you have the necessary
configuration established in this step, you should be able to do
future work by only doing steps
**2** ([starting a server](#2-start-the-jupyter-server)), and
**3** ([running your notebooks](#3-run-notebooks)) for your routine workflow.

This software environment refers to steps taken on the HPC host.

### 1.a) Download materials to HPC account

* Log in to the HPC machine
* Run this command:

```text

> curl -s -L 'https://github.com/USGS-python/hytest_notebook_tutorials/releases/latest/download/HyTEST-Tutorials.zip' | tar xvf -

Author: you are correct.  
The downloadable bundle needs to be formalized/defined....  That line was working on the idea that a subset of all of the material would be downloadable as a zip package. Not sure if that's the way you want to go.  Do we just want folks to clone the whole repo? 

```
git clone https://github.com/hytest-org/hytest.git
```
That's certainly easier, but likely a much bigger download, as it fetches the git history for the whole repo.  We can reduce that substantially by asking folks to clone with:
```
git clone --depth 1 https://github.com/hytest-org/hytest.git
```
I've not worked with the `depth` parameter much, so will need to test.

> curl -s -L 'https://github.com/USGS-python/hytest_notebook_tutorials/releases/latest/download/HyTEST-Tutorials.zip' | tar xvf -

Reviewer: Can we just download whichever files they need for now? (is it the yml and .sh scripts?). Some people are already on our Github poking around with CONUS404 data, so we need to make sure the content on our `main` branch works.

> curl -s -L 'https://github.com/USGS-python/hytest_notebook_tutorials/releases/latest/download/HyTEST-Tutorials.zip' | tar xvf -

Reviewer: I think cloning the whole repo might get unwieldy as we add more content

```

> **NOTE**: The above URL is a placeholder until we get set up with our release schedule and standards.  It is not likely to work yet.

You should now have a directory, `hytest` into which this archive has
been placed.

```text
> cd hytest
> ls -l
...
```

### 1.b) Set up a `hytest` Conda environment

> **NOTE**: We are using 'hytest' as the environment name.  If you need to use
another name, you'll need to make some adjustments to the

> [environment file](./HPC/HyTest.yml).

Reviewer: ```suggestion
[environment file](./environment_set_up/HyTest.yml).
```

> [environment file](./environment_set_up/HyTest.yml).

Reviewer: ```suggestion
[environment file](./environment_set_up/HyTEST.yml).
```

Conda is a package manager -- it will automate much of the software downloading
and configuring that we will need to run these notebooks. Use the `HyTEST.yml` environment definition file as the input specification to create a
new environment which includes the necessary software:

```text

> > conda env create -f ./environment_set_up/HyTest.yml

Reviewer: ```suggestion
> conda env create -f ./environment_set_up/HyTEST.yml
```

...
...
```

> **NOTE**: We're using the `conda` given to us from the sysadmin. You may
see other conda-like replacements used on other environments (`mamba` is a popular
one).  They are functionally equivalent for what we need to do here.

### 1.c) Configure Jupyter Server

The Jupyter Server is the element which will actually run the notebooks
which display in a remote (i.e. your desktop) web browser.  We have an auto-configuration script which will help with this configuration. Run:

```text

> > ./HPC/auto-conf.sh

Reviewer: ```suggestion
> ./environment_set_up/auto-conf.sh
```

Set the access password for connecting to this Jupyter Server...
Enter password:
Verify password:
...
```

The password you supply will be used when you attempt to connect to the
Jupyter server from your local web browser.

Note that this <kbd>auto-conf.py</kbd> script handles a lot of
configuration that you need only do once. It writes a configuration
file to `$HOME/.jupyter/jupyter_server_config.json`, which will simplify
the process of starting the server in future steps.

If you need (or want) to deviate from the vanilla set-up created by
`auto-conf.py` you can find more detailed information for setting up

> manually [here](./HPC/ManualConfig.md)

Reviewer: ```suggestion
manually [here](./environment_set_up/ManualConfig-HPC.md)
```

With that, your jupter server and its necessary environment should be set.

## 2) Start the jupyter server

Start Jupyter Server on an Interactive Compute Node

```text
> salloc -A account_name -t 02:00:00 -N 1 srun --pty bash
> conda activate hytest
> jupyter lab
```

<kbd>account_name</kbd> is your account credential/name

> **NOTE** If you chose to configure more [manually](./HPC/ManualConfig.md),

Reviewer: ```suggestion
**NOTE** If you chose to configure more [manually](./environment_set_up/ManualConfig-HPC.md),
```

# Quick-Start for HPC

This document will help you set up the correct computing environment on a compute
server (an HPC node), and access the notebooks on the HPC via your local desktop.

> 

Reviewer: ```suggestion

This document assumes that you already have an access to either the `Denali` or `Tallgrass` [supercomputers at USGS](https://hpcportal.cr.usgs.gov/hpc-user-docs/index.html) and are comfortable using the command line.

```

## 1) Configure the Software Environment

This should only need to be done once. After you have the necessary
configuration established in this step, you should be able to do
future work by only doing steps
**2** ([starting a server](#2-start-the-jupyter-server)), and
**3** ([running your notebooks](#3-run-notebooks)) for your routine workflow.

> This software environment refers to steps taken on the HPC host.

Reviewer: ```suggestion
This software environment refers to steps taken to configure the HPC host.
```

```text
> ./environment_set_up/auto-conf.py
Set the access password for connecting to this Jupyter Server...
Enter password:
Verify password:
...
```

The password you supply will be used when you attempt to connect to the
Jupyter server from your local web browser.

Note that this `auto-conf.py` script handles a lot of
configuration that you need only do once. It writes a configuration
file to `$HOME/.jupyter/jupyter_server_config.json`, which will simplify
the process of starting the server in future steps.

If you need (or want) to deviate from the vanilla set-up created by
`auto-conf.py` you can find more detailed information for setting up
manually [here](ManualConfig-HPC.md)

With that, your jupter server and its necessary environment should be set.

## 2) Start the Jupyter Server

The computation and data access will run on the compute node, and render
results via web connection to the browser on your desktop. This model
uses a 'server' process on the HPC in order to coordinate that connection.

Start Jupyter Server on an interactive compute node

```text
> salloc -A account_name -t 02:00:00 -N 1 srun --pty bash
> conda activate hytest
> jupyter lab
```

> `account_name` is your account credential/name

Reviewer: What account? is this my usgs email address?

Author: This is adapted from the text in here:
https://github.com/USGS-python/hytest-development/blob/main/hpc.md#installing-a-software-environment

it talks about needing to 'use your account name'... I presume this is the login name for getting access to the HPC server.

> **NOTE** If you chose to configure more [manually](./environment_set_up/ManualConfig-HPC.md),

Reviewer: ```suggestion
**NOTE** If you chose to configure more [manually](./ManualConfig-HPC.md),
```

you may need to supply extra options to that `jupyter lab` command, or use the
`start_jupyter.sh` example script we've provided for that purpose.

## 3) Run Notebooks

You are now ready to connect a web browser to your jupyter server.  Launch
a web browser on your dekstop and connect to the URL provided in the
startup messaging when you launched `jupyter lab`.

> NOTE:  Don't use the `localhost` option.  Use the URL which includes
your HPC host's full network name. That should look something like
<kbd>https://denali:8402/</kbd>.

The Jupyter Server will ask for a password.  This is the one supplied
in [Step 1c](#1c-configure-jupyter-server) above.

> You can now run existing notebooks found in `hytest` , or create

Reviewer: ```suggestion
You can now run existing notebooks found in `hytest`, or create
```

your own.

## 4) Shut Down Server

After a daily session, you will want to shut down the jupyter server.

> In theterminal session where you started `jupyter lab`, merely press Ctl-C

Reviewer: ```suggestion
In the terminal session where you started `jupyter lab`, merely press Ctl-C
```

```text
> ./environment_set_up/auto-conf.py
Set the access password for connecting to this Jupyter Server...
Enter password:
Verify password:
...
```

The password you supply will be used when you attempt to connect to the
Jupyter server from your local web browser.

Note that this `auto-conf.py` script handles a lot of
configuration that you need only do once. It writes a configuration
file to `$HOME/.jupyter/jupyter_server_config.json`, which will simplify
the process of starting the server in future steps.

If you need (or want) to deviate from the vanilla set-up created by
`auto-conf.py` you can find more detailed information for setting up
manually [here](ManualConfig-HPC.md)

With that, your jupter server and its necessary environment should be set.

## 2) Start the Jupyter Server

The computation and data access will run on the compute node, and render
results via web connection to the browser on your desktop. This model
uses a 'server' process on the HPC in order to coordinate that connection.

Start Jupyter Server on an interactive compute node

```text
> salloc -A account_name -t 02:00:00 -N 1 srun --pty bash
> conda activate hytest
> jupyter lab
```

> `account_name` is your account credential/name

Reviewer: ```suggestion
`account_name` is one of your accounts codes on the supercomputer. You will see a list of your account codes when you first log in to the supercomputer. They are listed in a table after some text saying `Your account codes are:`.
```
